crates/polars-core/src/series/ops/reshape.rs

```diff
@@ -89,31 +89,58 @@ impl Series {
 
     #[cfg(feature = "dtype-array")]
     pub fn reshape_array(&self, dimensions: &[i64]) -> PolarsResult<Series> {
+        if dimensions.is_empty() {
+            polars_bail!(InvalidOperation: "at least one dimension must be specified")
+        }
+
         let mut dims = dimensions.iter().copied().collect::<VecDeque<_>>();
 
         let leaf_array = self.get_leaf_array();
-        let size = leaf_array.len() as i64;
-
-        // Infer dimension
-        if dims.contains(&-1) {
-            let infer_dims = dims.iter().filter(|d| **d == -1).count();
-            polars_ensure!(infer_dims == 1, InvalidOperation: "can only infer single dimension, found {}", infer_dims);
-
-            let mut prod = 1;
-            for &dim in &dims {
-                if dim != -1 {
-                    prod *= dim;
+        let size = leaf_array.len();
+
+        let mut total_dim_size = 1;
+        let mut infer_dim_index: Option<usize> = None;
+        for (index, &dim) in dims.iter().enumerate() {
+            if dim > 0 {
+                total_dim_size *= dim as usize;
+            } else if dim == 0 {
+                if index != 0 {
+                    polars_bail!(
+                        InvalidOperation: "cannot reshape array into shape containing a zero dimension after the first: {}",
+                        format_tuple!(dims)
+                    );
                 }
+                total_dim_size = 0;
+                // We can early exit here, as empty arrays will error with multiple dimensions,
+                // and non-empty arrays will error when the first dimension is zero.
+                break;
+            } else if infer_dim_index.is_none() {
```

Comment:
nit: we could make this a single `else` branch and start the branch with a `polars_ensure(infer_index.is_none())`. This reads a bit easier and adds a `cold` hint to the compiler for the happy path.